internal/otelarrow/admission/boundedqueue.go

```diff
@@ -4,157 +4,174 @@
 	}
 }
 
-func (bq *BoundedQueue) admit(pendingBytes int64) (bool, error) {
+// acquireOrGetWaiter returns with three distinct conditions depending
+// on whether it was accepted, rejected, or asked to wait.
+//
+// - element=nil, error=nil: the fast success path
+// - element=nil, error=non-nil: the fast failure path
+// - element=non-nil, error=non-nil: the slow success path
+func (bq *BoundedQueue) acquireOrGetWaiter(pending uint64) (*list.Element, error) {
 	bq.lock.Lock()
 	defer bq.lock.Unlock()
 
-	if pendingBytes > bq.maxLimitBytes { // will never succeed
-		return false, fmt.Errorf("rejecting request, request size larger than configured limit")
+	if pending > bq.maxLimitAdmit {
+		// when the request will never succeed because it is
+		// individually over the total limit, fail fast.
+		return nil, ErrRequestTooLarge
 	}
 
-	if bq.currentBytes+pendingBytes <= bq.maxLimitBytes { // no need to wait to admit
-		bq.currentBytes += pendingBytes
-		return true, nil
+	if bq.currentAdmitted+pending <= bq.maxLimitAdmit {
+		// the fast success path.
+		bq.currentAdmitted += pending
+		return nil, nil
 	}
 
 	// since we were unable to admit, check if we can wait.
-	if bq.currentWaiters+1 > bq.maxLimitWaiters { // too many waiters
-		return false, ErrTooManyWaiters
+	if bq.currentWaiting+pending > bq.maxLimitWait {
+		return nil, ErrTooMuchWaiting
 	}
 
-	// if we got to this point we need to wait to acquire bytes, so update currentWaiters before releasing mutex.
-	bq.currentWaiters++
-	return false, nil
+	// otherwise we need to wait
+	return bq.addWaiterLocked(pending), nil
 }
 
-func (bq *BoundedQueue) Acquire(ctx context.Context, pendingBytes int64) error {
-	success, err := bq.admit(pendingBytes)
-	if err != nil || success {
-		return err
+// Acquire implements Queue.
+func (bq *BoundedQueue) Acquire(ctx context.Context, pending uint64) (ReleaseFunc, error) {
+	element, err := bq.acquireOrGetWaiter(pending)
+	parentSpan := trace.SpanFromContext(ctx)
+	pendingAttr := trace.WithAttributes(attribute.Int64("pending", int64(pending)))
+
+	if err != nil {
+		parentSpan.AddEvent("admission rejected (fast path)", pendingAttr)
+		return noopRelease, err
+	} else if element == nil {
+		parentSpan.AddEvent("admission accepted (fast path)", pendingAttr)
+		return bq.releaseFunc(pending), nil
 	}
 
-	// otherwise we need to wait for bytes to be released
-	curWaiter := waiter{
-		pendingBytes: pendingBytes,
-		readyCh:      make(chan struct{}),
-	}
+	parentSpan.AddEvent("enter admission queue")
 
-	bq.lock.Lock()
-
-	// generate unique key
-	for {
-		id := uuid.New()
-		_, keyExists := bq.waiters.Get(id)
-		if keyExists {
-			continue
-		}
-		bq.waiters.Set(id, curWaiter)
-		curWaiter.ID = id
-		break
-	}
-
-	bq.lock.Unlock()
-	ctx, span := bq.tracer.Start(ctx, "admission_blocked",
-		trace.WithAttributes(attribute.Int64("pending", pendingBytes)))
+	ctx, span := bq.tracer.Start(ctx, "admission_blocked", pendingAttr)
```

Comment:
Note: the use of `list.List` avoids the need to generate a new key. This simplification is possible because we always remove in LIFO order and don't require random access.